internals/cli/masker/stream.go

```diff
@@ -0,0 +1,122 @@
+package masker
+
+import (
+	"bytes"
+	"io"
+	"sync"
+)
+
+// stream is a buffered io.Writer that masks all secrets written on it using a multipleMatcher.
+type stream struct {
+	dest          io.Writer
+	buf           IndexedBuffer
+	registerFrame func(*stream, int)
+
+	matcher     *multipleMatcher
+	matches     Matches
+	matchesLock sync.Mutex
+}
+
+// Write implements the io.Writer interface for the stream.
+// The written frame is stored in the buffer and it is registered in the Masker to make sure it is flushed from
+// the buffer after the buffer delay has passed.
+// The bytes are also passed to the secret matcher to check for any matches with secrets.
+func (s *stream) Write(p []byte) (int, error) {
+	n, err := s.buf.Write(p)
+	if n > 0 {
+		s.registerFrame(s, n)
+	}
+
+	for index, length := range s.matcher.Write(p[:n]) {
+		s.addMatch(index, length)
+	}
+
+	return n, err
+}
+
+// addMatch adds the match of a secret at the given index and with the given length to the map of matches if the
+// associated bytes have not yet been written to the destination.
```

Comment:
Consider adding a tiny comment here why it is important that the bytes have not been written to the destination yet. We talked about this and there is actual important insight behind this I think is good to document for future readers of the code.